<a href="https://colab.research.google.com/github/RaphaelDdS/Alura_pythonDataScience/blob/main/1PrimeirosPassos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print('Escola de Dados da ALura!')

Escola de Dados da ALura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
nome = 'Raphael'
sobrenome = 'Domingues'

print(f'Nome: {nome} \nSobrenome: {sobrenome}')

Nome: Raphael 
Sobrenome: Domingues


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [ ]:
print(f'R \nA \nP \nH \nA \nE \nL')

R 
A 
P 
H 
A 
E 
L


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
dia = 9
mes = 'Julho'
ano = 1996

print(dia, mes, ano)

9 Julho 1996


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
ano_atual = 2024
print(f'Ano atual: {ano_atual}')

Ano atual: 2024


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
nome = input('Digite seu nome: ')
print(f'Olá, {nome}!')

Digite seu nome: Raphael
Olá, Raphael!


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
nome = input('Digite seu nome: ')
idade = input('Digite sua idade: ')
print(f'Olá, {nome}! Você tem {idade} anos.')

Digite seu nome: Raphael
Digite sua idade: 27
Olá, Raphael! Você tem 27 anos.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
nome = input('Insira seu nome: ')
idade = input('Insira sua idade: ')
altura = input('Insira sua altura em metros: ')

print(f'Olá, {nome}, você tem {idade} anos e mede {altura} metros!')

Insira seu nome: Raphael
Insira sua idade: 27
Insira sua altura em metros: 1.80
Olá, Raphael, você tem 27 anos e mede 1.80 metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
x = int(input('Insira primeiro valor: '))
y = int(input('Insira segundo valor: '))
total = x + y

print(f'A soma de {x} + {y} é: {total}!')

Insira primeiro valor: 9
Insira segundo valor: 49
A soma de 9 + 49 é: 58!


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
x = int(input('Insira o primeiro valor: '))
y = int(input('Insira o segundo valor: '))
z = int(input('Insira o terceiro valor: '))

soma = x + y + z

print(f'Os valores inseridos foram: {x}, {y} e {z}. A soma é: {soma}')

Insira o primeiro valor: 2
Insira o segundo valor: 47
Insira o terceiro valor: 3723
Os valores inseridos foram: 2, 47 e 3723. A soma é: 3772


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
x = int(input('Insira o primeiro valor: '))
y = int(input('Insira o segundo valor: '))
total = x - y

print(f'Os valores inseridos foram: {x} e {y}. A subtração do primeiro pelo segundo valor é: {total}!')

Insira o primeiro valor: 24
Insira o segundo valor: 3
Os valores inseridos foram: 24 e 3. A subtração do primeiro pelo segundo valor é: 21!


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
x = int(input('Insira o primeiro valor: \n'))
y = int(input('Insira o segundo valor: \n'))
total = x * y

print(f'Os valores inseridos foram: {x} e {y}! \nA multiplicação dos dois valores é: {total}!')

Insira o primeiro valor: 
3
Insira o segundo valor: 
6
Os valores inseridos foram: 3 e 6! 
A multiplicação dos dois valores é: 18!


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
x = int(input('Insira o primeiro valor (NUMERADOR): \n'))
y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): \n'))

while y == 0:
  y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): \n'))
  if y != 0:
    break

total = x / y

print(f'Os números inseridos foram: {x} e {y}! \nA divisão entre os dois valores é: {total}!')

Insira o primeiro valor (NUMERADOR): 
6
Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): 
2
Os números inseridos foram: 6 e 2! 
A divisão entre os dois valores é: 3.0!


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
x = int(input('Insira o primeiro valor (OPERADOR): \n'))
y = int(input('Insira o segundo valor (POTÊNCIA): \n'))
total = x ** y

print(f'Os valores inseridos foram: {x} e {y}! \nA exponenciação entre os dois valores é: {total}')



Insira o primeiro valor (OPERADOR): 
4
Insira o segundo valor (POTÊNCIA): 
6
Os valores inseridos foram: 4 e 6! 
A exponenciação entre os dois valores é: 4096


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
x = int(input('Insira o primeiro valor (NUMERADOR): \n'))
y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): \n'))

while y == 0:
  y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): \n'))
  if y != 0:
    break

total = x // y

print(f'Os valores inseridos foram: {x} e {y}! \nA divisão INTEIRA entre os dois valores é: {total}!')

Insira o primeiro valor (NUMERADOR): 
29
Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): 
4
Os valores inseridos foram: 29 e 4! 
A divisão INTEIRA entre os dois valores é: 7!


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
x = int(input('Insira o primeiro valor (NUMERADOR): \n'))
y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): \n'))

while y == 0:
  y = int(input('Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO: \n) '))
  if y != 0:
    break

total = x % y

print(f'Os valores inseridos foram: {x} e {y}! \nO RESTO da divisão entre os dois valores é: {total}!')

Insira o primeiro valor (NUMERADOR): 
392
Insira o segundo valor (DENOMINADOR, DIFERENTE DE ZERO): 
3
Os valores inseridos foram: 392 e 3! 
O RESTO da divisão entre os dois valores é: 2!


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
x = int(input('Insira a primeira nota: \n'))
y = int(input('Insira a segunda nota: '))
z = int(input('Insira a terceira nota: '))

media = (x + y + z) / 3

print(f'As notas inseridas foram: {x}, {y} e {z}! \nA média é: {media}!')

Insira a primeira nota: 
9
Insira a segunda nota: 3
Insira a terceira nota: 8
As notas inseridas foram: 9, 3 e 8! 
A média é: 6.666666666666667!


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
v1 = 5
v2 = 12
v3 = 20
v4 = 15

p1 = 1
p2 = 2
p3 = 3
p4 = 4

soma_ponderada = ((v1*p1)+(v2*p2)+(v3*p3)+(v4*p4))
n = p1 + p2 + p3 + p4
media_ponderada = soma_ponderada / n

print(f"A média ponderada é: {media_ponderada}!")

A média ponderada é: 14.9!


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase = 'Olha que frase legal!'
print(frase)

Olha que frase legal!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frase = input('Digite uma frase: \n')
print(frase)

Digite uma frase: 
Olha essa frase!
Olha essa frase!


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
frase = input('Digite uma frase: \n')


print(frase.upper())

Digite uma frase: 
aimeudeus
AIMEUDEUS


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
frase = input('Digite uma frase: \n')

print(frase.lower())

Digite uma frase: 
AIMEUDEUS
aimeudeus


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase = ' String com espaço na frente e atrás '

print(frase.strip())

String com espaço na frente e atrás


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase = input('Digite uma frase: \n')

print(frase.strip())

Digite uma frase: 
     aiaiaia       
aiaiaia


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase = input('Digite uma frase: \n')

print(frase.strip().lower())

Digite uma frase: 
      AIMEUDEUS  
aimeudeus


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input('Digite uma frase: \n')

print(frase.replace('e','f'))

Digite uma frase: 
Olha só essa frase!
Olha só fssa frasf!


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input("Digite uma frase: \n")

frase = frase.replace("a", chr(64))

print(frase)


Digite uma frase: 
Olha só essa frase!
Olh@ só ess@ fr@se!


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input("Digite uma frase: \n")

frase = frase.replace("s", chr(36))

print(frase)

Digite uma frase: 
Olha só essa frase!
Olha $ó e$$a fra$e!


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
v1 = int(input('Insira o primeiro valor: \n'))
v2 = int(input('Insira o segundo valor: \n'))

if v1 > v2:
  print(f'O primeiro valor ({v1}) é maior que o segundo valor ({v2})!')
else:
  print(f'O segundo valor ({v2}) é maior que o primeiro valor ({v1})!')

Insira o primeiro valor: 
7
Insira o segundo valor: 
49
O segundo valor (49) é maior que o primeiro valor (7)!


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
percentual = float(input('Insira o percentual de crescimento: \n'))

if percentual > 0:
  print(f'A empresa teve um CRESCIMENTO de {percentual}{chr(37)}!')
elif percentual < 0:
  print(f'A empresa teve um DECRESCIMENTO de {percentual}{chr(37)}!')
else:
  print(f'A empresa estagnou. :(')

Insira o percentual de crescimento: 
0
A empresa estagnou. :(


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
vogais = 'a, e, i, o, u'

letra = input('Insira uma letra: \n').lower()

if letra in vogais:
  print(f'A letra {letra} é uma vogal!')
else:
  print(f'A letra {letra} é uma consoante!')

Insira uma letra: 
d
A letra d é uma consoante!


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
v1 = float(input('Valor médio ano 1: \n'))
v2 = float(input('Valor médio do ano 2: \n'))
v3 = float(input('Valor médio do ano 3: \n'))

if v1 > v2 and v1 > v3:
  v_maior = v1
elif v1 < v2 and v1 < v3:
  v_menor = v1
else:
  v_meio = v1

if v2 > v1 and v2 > v3:
  v_maior = v2
elif v2 < v1 and v2 < v3:
  v_menor = v2
else:
  v_meio = v2

if v3 > v1 and v3 > v2:
  v_maior = v3
elif v3 < v1 and v3 < v2:
  v_menor = v3
else:
  v_meio = v3

print(f'O valor mais alto foi: {v_maior}! \nO valor mais baixo foi: {v_menor}! \nConsequentemente, o vamor mediano foi: {v_meio}!')


Valor médio ano 1: 
5
Valor médio do ano 2: 
6
Valor médio do ano 3: 
7
O valor mais alto foi: 7.0! 
O valor mais baixo foi: 5.0! 
Consequentemente, o vamor mediano foi: 6.0!


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
p1 = "Banana"
p2 = "Maçã"
p3 = "Chevrolet 2002"

v1 = float(input(f'Insira o valor do produto "{p1}": \n'))
v2 = float(input(f'Insira o valor do produto "{p2}": \n'))
v3 = float(input(f'Insira o valor do produto "{p3}": \n'))

if v1 < v2 and v1 < v3:
  print(p1, v1)
elif v2 < v1 and v2 < v3:
  print(p2, v2)
else:
  print(p3, v3)

Insira o valor do produto "Banana": 
896
Insira o valor do produto "Maçã": 
3
Insira o valor do produto "Chevrolet 2002": 
244
Maçã 3.0


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
n1 = int(input('Insira o primeiro número: \n'))
n2 = int(input('Insira o segundo número: \n'))
n3 = int(input('Insira o terceiro número: \n'))

if n1 > n2 and n1 > n3:
  maior = n1
elif n1 < n2 and n1 < n3:
  menor = n1
else:
  meio = n1

if n2 > n1 and n2 > n3:
  maior = n2
elif n2 < n1 and n2 < n3:
  menor = n2
else:
    meio = n2

if n3 > n1 and n3 > n2:
  maior = n3
elif n3 < n1 and n3 < n2:
  menor = n3
else:
  meio = n3

print(maior, meio, menor)

Insira o primeiro número: 
4
Insira o segundo número: 
239
Insira o terceiro número: 
210581258
210581258 239 4


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [ ]:
turno = input('Insira o período em que você estuda (manhã, tarde ou noite): \n').lower()
turnos = ['manhã', 'tarde', 'noite']

while turno not in turnos:
  print('Valor inválido!')
  turno = input('Insira o período em que você estuda (manhã, tarde ou noite): \n').lower()

if turno == "manhã":
  print('Bom dia!')
elif turno == "tarde":
  print('Boa tarde!')
else:
  print('Boa noite!')


Insira o período em que você estuda (manhã, tarde ou noite): 
MANHÃ
Bom dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
n = int(input('Insira um número: \n'))
if n % 2 == 0:
  print(f'O número {n} é par!')
else:
  print(f'O número {n} é impar!')

Insira um número: 
6
O número 6 é par!


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
n = input('Insira um número: \n')

if n.isdigit():
  print(f'O número {n} é inteiro!')
else:
  print(f'O número {n} é decimal!')


Insira um número: 
2.4
O número 2.4 é decimal!


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [ ]:
n1 = int(input('Insira um número: '))
n2 = int(input('Insira outro número: \n'))
operador = input('Insira o operador (+, -, /, *): ')

if operador == '+':
    total = n1 + n2
elif operador == '-':
    total = n1 - n2
elif operador == '*':
    total = n1 * n2
elif operador == '/':
    if n2 != 0:
        total = n1 / n2
    else:
        total = 'Erro. Divisão por zero!'
else:
    total = 'Operador inválido.'

if total % 2 != 0:
  paridade = "ímpar"
else:
  paridade = "par"

if total > 0:
  positividade = "positivo"
elif total < 0:
  positividade = "negativo"
else:
  positividade = "neutro"

if type(total) == float:
  interidade = 'decimal'
else:
  interidade = 'inteiro'

print(f'O resultado de {n1} {operador} {n2} é: {total}')
print(f'{total} é um número {paridade}, {positividade} e {interidade}')


Insira um número: 83
Insira outro número: 
25
Insira o operador (+, -, /, *): /
O resultado de 83 / 25 é: 3.32
3.32 é um número ímpar, positivo e decimal


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [ ]:
l1 = int(input('Primeiro lado: '))
l2 = int(input('Segundo lado: '))
l3 = int(input('Terceiro lado: '))

if l1 + l2 > l3 or l1 + l3 > l2 or l2 + l3 > l1:
  triangulo = 'triângulo'
else:
  triangulo = "não triângulo"

if triangulo == 'triângulo':
  if l1 == l2 and l2 == l3:
    triangulo_tipo = 'equilátero'
  elif l1 == l2 or l2 == l3 or l1 == l3:
   triangulo_tipo = 'isóceles'
  else:
    triangulo_tipo = 'escaleno'
else:
  triangulo_tipo = 'de nenhum tipo'

print(f'Lado 1 \t {l1} \nLado 2 \t {l2} \nLado 3 \t {l3}')
print(f'Os tres lados formam um {triangulo} {triangulo_tipo}!')

Primeiro lado: 9
Segundo lado: 8
Terceiro lado: 2
Lado 1 	 9 
Lado 2 	 8 
Lado 3 	 2
Os tres lados formam um triângulo escaleno!


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [ ]:
d_preco = 2.00
e_preco = 1.70

combustivel = input('Qual combustível gostaria de comprar? Insira "D" para Diesel e "E" para Etanol: \n').upper()
if combustivel != "E" and combustivel != "D":
  print('Combustível inválido!')
  exit()
litros = int(input('Quantos litros deseja comprar? \nReforçando que acima de 15 litros você ganha mais desconto!\n'))

if combustivel == "E":
  if litros > 15:
    p_desconto = 0.04
    desconto = round(((e_preco * litros) * p_desconto), 2)
    total = (e_preco * litros) - desconto
  else:
    p_desconto = 0.02
    desconto = round(((e_preco * litros) * p_desconto), 2)
    total = (e_preco * litros) - desconto
elif combustivel == "D":
  if litros > 15:
    p_desconto = 0.05
    desconto = round(((d_preco * litros) * p_desconto), 2)
    total = (d_preco * litros) - desconto
  else:
    p_desconto = 0.03
    desconto = round(((d_preco * litros) * p_desconto), 2)
    total = (d_preco * litros) - desconto
else:
  print('Combustível inválido!')
print(f'Seu abastecimento custou {total}R{chr(36)}. \nVocê teve um desconto de {desconto}R{chr(36)}!')

Combustível inválido!


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
